flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java

```diff
@@ -0,0 +1,258 @@
+import org.junit.rules.TestName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+import static com.netease.arctic.flink.shuffle.RowKindUtil.transformFromFlinkRowKind;
+import static com.netease.arctic.flink.table.descriptors.ArcticValidator.ARCTIC_LOG_CONSISTENCY_GUARANTEE_ENABLE;
+import static com.netease.arctic.flink.util.kafka.KafkaContainerTest.KAFKA_CONTAINER;
+import static com.netease.arctic.flink.util.kafka.KafkaContainerTest.getPropertiesByTopic;
+import static com.netease.arctic.flink.util.kafka.KafkaContainerTest.readRecordsBytes;
+import static com.netease.arctic.flink.write.hidden.TestBaseLog.createLogDataDeserialization;
+import static com.netease.arctic.flink.write.hidden.TestBaseLog.userSchema;
+import static com.netease.arctic.flink.write.hidden.TestHiddenLogOperators.createRowData;
+import static org.junit.Assert.assertEquals;
+
+public class TestKafkaSourceReader {
+  private static final Logger LOG = LoggerFactory.getLogger(TestKafkaSourceReader.class);
+  private static String topic;
+  private static final int NUM_SPLITS = 1;
+  private static final int NUM_RECORDS_PER_SPLIT = 10;
+  private static final int TOTAL_NUM_RECORDS = NUM_RECORDS_PER_SPLIT * NUM_SPLITS;
+
+  @Rule
+  public TestName testName = new TestName();
+
+  private static final byte[] JOB_ID = IdGenerator.generateUpstreamId();
+
+  @BeforeClass
+  public static void prepare() throws Exception {
+    KAFKA_CONTAINER.start();
+  }
+
+  @AfterClass
+  public static void shutdown() throws Exception {
+    KAFKA_CONTAINER.close();
+  }
+
+  @Before
+  public void initData() throws Exception {
+    topic = TestUtil.getUtMethodName(testName);
+    write(topic, TOTAL_NUM_RECORDS);
+  }
+
+  @Test
+  public void testSourceReaderFailover() throws Exception {
+    final String groupId = "testSourceReaderFailover";
+    LogKafkaSourceReader reader = (LogKafkaSourceReader) createReader(groupId);
+    reader.addSplits(getSplits(NUM_SPLITS));
+    ValidatingSourceOutput output = new ValidatingSourceOutput();
+    List<KafkaPartitionSplit> splitList;
+    long checkpointId = 0;
+    do {
+      checkpointId++;
+      reader.pollNext(output);
+      // Create a checkpoint for each message consumption, but not complete them.
+      splitList = reader.snapshotState(checkpointId);
+    } while (output.count() < TOTAL_NUM_RECORDS);
+
+    // The completion of the last checkpoint should subsume all the previous checkpoints.
+    assertEquals(checkpointId, reader.getOffsetsToCommit().size());
+    reader.notifyCheckpointComplete(checkpointId);
+
+    // re-create and restore
+    reader = (LogKafkaSourceReader) createReader(groupId);
+    reader.addSplits(splitList);
+    List<KafkaPartitionSplit> currentSplitList = reader.snapshotState(checkpointId++);
```

Comment:
Why `checkpoint++` here?  I think only `checkpoint`, WDYT?

Reply:
Yes, you're right, I've modified it.